### ISSUE TITLE
[BUG] Incorrect backward in `_ScatterToSequenceParallelRegion`: should not use `gather`

### ISSUE THREAD

**Issue by qinxiangyujiayou** [NVIDIA/Megatron-LM#1635]
### **Describe the bug**

In the **`_ScatterToSequenceParallelRegion`**, the current implementation of the `backward` function incorrectly uses `_gather_along_first_dim`, which breaks the correctness of **sequence parallelism** gradient flow.

```python
@staticmethod
def backward(ctx, grad_output):
    return _gather_along_first_dim(grad_output, ctx.group), None
```

This gathers **all gradients from all ranks**, even though **only one rank** is responsible for its corresponding slice. This results in **duplicate and incorrect gradients** during backpropagation.

---

### **Why it’s incorrect**

In the **forward** pass, each rank gets only a **slice of the input tensor** along dimension 0:

```python
X_total -> split along dim 0 -> X_rank_i
```

In the **backward** pass, each rank only computes gradients with respect to **its local slice**, so the gradient with respect to the full input should be:

* Zero everywhere **except** in the rank’s own slice.
* Not a gather of all gradients (which would duplicate gradient contributions).

---

### **Expected behavior**

Each rank should return a zero-initialized gradient tensor, with **only its assigned slice** filled from `grad_output`, like this:

```python
@staticmethod
def backward(ctx, grad_output):
    rank = torch.distributed.get_rank(ctx.group)
    world_size = torch.distributed.get_world_size(ctx.group)
    input_shape = ctx.input_shape  # should be saved in forward()

    chunk_size = input_shape[0] // world_size
    start = rank * chunk_size
    end = start + grad_output.shape[0]

    grad_input = torch.zeros(input_shape, dtype=grad_output.dtype, device=grad_output.device)
    grad_input[start:end] = grad_output

    return grad_input, None
```

---

### **Proposed Fix**

* **Store** the original input shape in `ctx` during `forward`
* **Replace** the `backward` implementation with a zero-padded slice write (as above)
* **Do not use `gather`**; it’s semantically incorrect in this context

---

### **Impact**

* ❌ **Incorrect gradients** in multi-GPU training with sequence parallelism
* ❌ **Gradient overcounting** across ranks
* ⚠️ **Potential training instability or divergence**

---

### **Optional: Proposed PR available**

Let me know if you'd like a PR with the fix — happy to contribute.

---
